### ISSUE TITLE
Design: Transaction Integrations API

### ISSUE THREAD

**Issue by msfeldstein** [stellar/django-polaris#15]
We need to create an API that will work for a wide range of banking providers so that any anchor can plug their bank rails into polaris.  This task is to create some documentation on an API and get buy-in from SDF and Anchor partners.

**Comment by msfeldstein** [comment]
Designed and implemented in https://github.com/stellar/django-polaris/pull/13

Docs viewable: https://build-pyqprdnkgw.now.sh/html/integrations/

**Comment by msfeldstein** [comment]
```
execute_deposit(transaction)
    You may override this function to fit your needs, but you must make a call to the create_stellar_deposit() CLI tool and update the status of the transaction as implemented below.
```

If there's some code that must always happen, would it be better to add before/after_execute_deposit() callbacks instead, and not require the implementor to include any code, only adding their own customization?

**Comment by msfeldstein** [comment]
```
 classmethod match_transaction(response, transaction)

    Determines whether or not the given response represents the given transaction...    You may override this function if you have another way of uniquely identifying a transaction.
```

What would be a different way of identifying the transaction, it seems pretty standard?

**Comment by msfeldstein** [comment]
> response (Dict) – a response body returned from Horizon for the transaction

Can we clarify what is in this Dict?  At least link to the horizon docs?

**Comment by msfeldstein** [comment]
```
update_transaction(was_processed, response, transaction)

    Updates the transaction depending on whether or not the transaction was successfully executed on the Stellar network and process_withdrawal raised an exception.

    If an exception was raised during process_withdrawal, we mark the corresponding Transaction status as error. If the Stellar transaction succeeded, we mark it as completed. Else, we mark it pending_stellar, so the wallet knows to resubmit.
```

I'm not sure i understand this.  It's a withdrawal so if the stellar transaction succeeded, but `process_withdrawal` had an error, it should be in error state right?  It's not completed if we didn't process the withdrawal

**Comment by JakeUrban** [comment]
> It's a withdrawal so if the stellar transaction succeeded, but `process_withdrawal` had an error, it should be in error state right? It's not completed if we didn't process the withdrawal

Thats correct. If `process_withdrawal` had an error, we'd mark the transaction status as `error`. Only when `process_withdrawal` completed successfully do we check if the Stellar transaction succeeded.

```
        if not was_processed:
            transaction.status = Transaction.STATUS.error
        elif response["successful"]:
            transaction.completed_at = now()
            transaction.status = Transaction.STATUS.completed
            transaction.status_eta = 0
            transaction.amount_out = transaction.amount_in - transaction.amount_fee
        else:
            transaction.status = Transaction.STATUS.pending_stellar
```

**Comment by JakeUrban** [comment]
> What would be a different way of identifying the transaction, it seems pretty standard?

I agree using the `withdraw_memo` to identify the transaction should be considered standard, so maybe we shouldn't make this function overridable. Especially since we don't allow the developer to control _how_ the memo field is constructed. I can take this out of the integrations class.

**Comment by JakeUrban** [comment]
> If there's some code that must always happen, would it be better to add before/after_execute_deposit() callbacks instead, and not require the implementor to include any code, only adding their own customization?

In general, I leaned on the side of _giving the developer more control_ when defining the API. Thats why I made `match_transaction`, `update_transaction`, and `execute_deposit` overridable even though we have default implementations. But you bring up a good point with `match_transaction`. If we believe developers wouldn't have a reason to change the functionality, then we probably shouldn't give them the option to do so.

For `execute_deposit` though, I can definitely see a reason to want to do some custom post-processing after a transaction has been submitted. I'll add an `after_execute_deposit` function for that, and remove `execute_deposit` as a overridable function.
